pkg/applications/providers/source/git.go

```diff
@@ -0,0 +1,217 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
+	gossh "golang.org/x/crypto/ssh"
+
+	appskubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/apps.kubermatic/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GitSource download the application's source from a git repository.
+type GitSource struct {
+	Ctx context.Context
+
+	// Client to seed cluster.
+	Client ctrlruntimeclient.Client
+
+	Source *appskubermaticv1.GitSource
+
+	// Namespace where credential secrets are stored.
+	SecretNamespace string
+}
+
+// DownloadSource clone the repository into destination and return the full path to the application's sources.
+func (g GitSource) DownloadSource(destination string) (string, error) {
+	auth, err := g.authFromCredentials()
+	if err != nil {
+		return "", err
+	}
+
+	checkout := g.getCheckoutStrategy()
+	if err := checkout(g.Ctx, destination, g.Source, auth); err != nil {
+		return "", fmt.Errorf("failed to clone repository: %w", err)
+	}
+
+	return path.Join(destination, g.Source.Path), nil
+}
+
+// authFromCredentials returns a new transport.AuthMethod according to credentials defined in the GitSource.
+// If no crednetials are defined in the GitSource then nil is returned.
+func (g GitSource) authFromCredentials() (transport.AuthMethod, error) {
+	var auth transport.AuthMethod
+	credentials := g.Source.Credentials
+	if credentials != nil {
+		switch credentials.Method {
+		case appskubermaticv1.GitAuthMethodPassword:
+			username, err := g.getCredentialFromSecret(credentials.Username.Name, credentials.Username.Key)
+			if err != nil {
+				return nil, err
+			}
+
+			password, err := g.getCredentialFromSecret(credentials.Password.Name, credentials.Password.Key)
+			if err != nil {
+				return nil, err
+			}
+
+			auth = &http.BasicAuth{Username: username, Password: password}
+
+		case appskubermaticv1.GitAuthMethodToken:
+			token, err := g.getCredentialFromSecret(credentials.Token.Name, credentials.Token.Key)
+			if err != nil {
+				return nil, err
+			}
+
+			auth = &http.TokenAuth{Token: token}
+		case appskubermaticv1.GitAuthMethodSSHKey:
+			privateKey, err := g.getCredentialFromSecret(credentials.SSHKey.Name, credentials.SSHKey.Key)
+			if err != nil {
+				return nil, err
+			}
+
+			authssh, err := ssh.NewPublicKeys("git", []byte(privateKey), "")
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse private ssh key: %w", err)
+			}
+			authssh.HostKeyCallback = gossh.InsecureIgnoreHostKey()
```

Comment:
I guess for now it is okay to allow any host. Maybe we should add a small section to the explanation of the func, that for now all hosts are accepted by default